Implement strftime and strptime for Time

vn-ki opened the pull request:
There is no easy formatting available for `Time`. I wanted to get some feedback on how to go about it.

This is my first time contributing here. Sorry if I did something wrong.

Closes #7309.

vn-ki commented:
Can I get some suggestions in regards to this PR? Is this the right way to go?

taldcroft commented:
Note prior discussion in #7309.

@vn-ki - thanks, this is a possible start.  As @mhvk noted there is a question about API and in particular behavior for arrays.

In looking at this I sort of think that the most clear would be to just add a `Time.strftime(self, format_spec)` method that essentially mimics the corresponding `datetime.strftime()` method.

Using `__format__` is fine, but it suffers from using a syntax language that is completely different from normal Python formatting and is really specific to `strftime`.  So using instead a Time `strftime` would  make user code more clear e.g. `tm.strftime('%Y-%m-%d %H:%M:%S')`.  That will also leave the `__format__` method free for us in the future to do something sane with it.

Thoughts @mhvk?

Also, arrays need to be handled.  I'd suggest if the input Time object is an array then the output should be a like-dimensioned string array as well.

Finally, as a follow-on PR you might add a `Time.strptime()` class method for symmetry, though of course the work-around is really simple since `Time` will accept a `datetime` object.

taldcroft commented:
@mhvk - I'm not sure I follow your concern about maintenance for a `strptime` or `strftime`.  They are quite independent of core functionality and won't add to overall complexity, e.g. this cannot be said of time masking :wink:.

At first glance these seem like a straightforward helper methods that at least SunPy would find useful.  Indeed the implementations will be pretty simple, but still decent-sized methods when you do it properly and support arrays, validation, error handling, so you don't want this code being rewritten everywhere.

mhvk commented:
@taldcroft - it is partially that I'd rather have it done right, but you pointed out the problems with `__format__` quite well. Anyway, if it is really helpful to SunPy, that is a good reason to have it, though possibly the right way to proceed would then be to try a subclass in SunPy and move it up. Anyway, since I don't really have time for this, I'll unsubscribe here.

taldcroft commented:
@vn-ki - BTW, does SunPy use many arbitrary format strings, or just a few such strings over and over?  If you don't need the ability to use arbitrary new format strings it could make sense to use the existing machinery with a new `TimeString` subclass that has the ones SunPy uses.

vn-ki commented:
Some of it is a configurable custom format. But there are quite some cases where we have to format the time to do a network request. For these, the format strings are different from one another.

We had discussed subclassing over at SunPy and since we had a lot of strftime, it was @cadair 's suggestion to add a function similar to strftime to astropy.

taldcroft commented:
@vn-ki - OK, then press forward with a `strftime` method on `Time`.

vn-ki commented:
@taldcroft What more do I need to do in this PR. Any changes required in the design or functionality?

Also I know I should add a doc string and change log entry. Other than that is there anything? Should I document the function somewhere?

vn-ki commented:
@taldcroft I decided to add `strptime` with this PR. Also I have added the tests and doc strings too.

Can I get a review from someone now?

vn-ki commented:
Thank you. Please take your time.

taldcroft commented:
I'm not an expert on `time` and `datetime`, but I think the code for `strftime` can be much simpler.  Hint:
```
In [15]: t = Time([[1,2],[3,4]], format='cxcsec')

In [16]: for dt in t.datetime.flat:
    ...:     print(dt.strftime('%Y %S'))
    ...:     
1998 01
1998 02
1998 03
1998 04
```
Other requirements:
- For `strptime` it must also accept a numpy str array or any Python list structure (or list of lists, etc) that can be converted to an N-d numpy array.
- For `strftime` in the case of an array it must output a numpy array with `shape` that corresponds to `self.shape`.

vn-ki commented:
@taldcroft About your first concern,
```python
>>> t = Time(['2016-12-31 23:59:60', '2015-06-30 23:59:60'])
>>> for dt in t.datetime.flat:
...     print(dt.strftime('%Y %S'))
...
ValueError: Time (array(2016, dtype=int32), array(12, dtype=int32), array(31, dtype=int32), array(23, dtype=int32), array(59, dtype=int32), array(60, dtype=int32), array(0, dtype=int32)) is within a leap second but datetime does not support leap seconds
>>>
>>> t.strftime('%Y %S')
['2016 60', '2015 60']
```

I will update the PR to satisfy other requirements.

taldcroft commented:
@vn-ki - ah, right.  I'm not sure offhand how to simplify, but I would guess there is a way.  That's your goal, to find an elegant way to solve this!

vn-ki commented:
@taldcroft I have changed `strftime` a bit. See if this is okay?

I have implemented your other suggestions.
Also, how can we test the equality of 2 `Time` objects which have more than 1 element? Do we have to iterate through each one?

taldcroft commented on astropy/time/core.py:
```diff
@@ -464,6 +499,49 @@ def __repr__(self):
+
+        """
+
+        if self.format == 'datetime':
```
I don't think you want to handle `datetime` specially like this because you can (in theory) have a `Time` object with `format='datetime'` which is within a leap second.  Just because the format is `datetime` does not mean that the `value` has been computed.  This is because you can change the format of an existing Time object.

taldcroft commented on astropy/time/core.py:
```diff
@@ -464,6 +499,49 @@ def __repr__(self):
+            info = list(self.replicate('iso')._time.str_kwargs())
+
+        else:
+            info = list(self._time.str_kwargs())
```
Delete everything from `if self.format == 'datetime'` to here, not needed.

taldcroft commented on astropy/time/core.py:
```diff
@@ -464,6 +499,49 @@ def __repr__(self):
+            info = list(self._time.str_kwargs())
+
+        formatted_strings = []
+        for i in info:
```
Untested, but i think this should work in all cases.  Note that `str_kwargs()` is a generator that automatically flattens the object and works for both scalar and array values.  This avoids creation of a temporary array.  Note also that single letter variables like `i` should generally be avoided, with the possible exception of within a list comprehension.
```
for sk in self.replicate('iso')._time.str_kwargs():
    date_tuple = date(sk['year'], sk['mon'], sk['day']).timetuple()
    .. etc

if self.isscalar:
    out = formatted_strings[0]
else:
    out = np.array(formatted_strings).reshape(self.shape)

return out
```

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
+
+        for v, s in iterator:
+            time_tuple = strptime(to_string(v), format_string)
+            s[...] = strftime('%Y-%m-%dT%H:%M:%S', time_tuple)
```
If you have the `time_tuple` can you just use standard Python string formatting?  Might this be faster?

taldcroft commented:
Tests appear to be failing.

vn-ki commented:
@taldcroft I can't seem to understand the test failure. (Travis shows seg fault)

pllim commented:
segfault is a known issue and is being investigated upstream. Maybe @bsipocz has update on this?

vn-ki commented:
@taldcroft I have made all the requested changes.

Cadair commented:
@vn-ki rebase or merge in master and the ci should fix I think.

vn-ki commented:
What happened :question:

taldcroft commented on astropy/time/core.py:
```diff
@@ -464,6 +499,35 @@ def __repr__(self):
+
+        """
+        formatted_strings = []
+        for sk in list(self.replicate('iso')._time.str_kwargs()):
```
Is this use of `list` necessary for some reason?  I think it should work fine without it and save creation of a temporary list of the whole time object.

taldcroft commented on astropy/time/core.py:
```diff
@@ -464,6 +499,35 @@ def __repr__(self):
+                              date_tuple[6], date_tuple[7], -1)
+            formatted_strings.append(strftime(format_spec, datetime_tuple))
+
+        if len(formatted_strings) == 1:
```
Use `self.isscalar` here to test for a scalar.

taldcroft commented on astropy/time/core.py:
```diff
@@ -464,6 +499,35 @@ def __repr__(self):
 
+    def strftime(self, format_spec):
+        """
+        Convert Time to a string according to a format specification.
```
This function converts to a string or numpy.array of strings.  Say so here and in the `Returns` section.

taldcroft commented on astropy/time/core.py:
```diff
@@ -464,6 +499,35 @@ def __repr__(self):
+    def strftime(self, format_spec):
+        """
+        Convert Time to a string according to a format specification.
+        See time.stftime documentation for format specification.
```
Should be `time.strftime`.  You should make this a link, which I think can be done with a Sphinx intersphinx link like:
```
See `time.strftime` documentation ..
```

vn-ki replied:
I should link the python documentation here?

bsipocz replied:
we're using the python intersphinx, so these will automatically turned into links. `time` is part of the standard library, so I think this will work as Tom suggested above (if you get build errors, let me know and I can have a more detailed look at it)

taldcroft commented on astropy/time/tests/test_basic.py:
```diff
@@ -1308,3 +1308,57 @@ def test_setitem_deltas():
+
+
+def test_strftime_array():
+    tstrings = ['2010-09-03 00:00:00', '2005-09-03 06:00:00',
```
Test with a date that is within a leap second.

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
+        Parse a string to a Time according to a format specification.
+        See time.strptime documentation for format specification.
+
+        >>> Time.strptime('2012-06-30T23:59:60', '%Y-%m-%dT%H:%M:%S')
```
Use an example date format that is not already supported by `Time`, e.g. `'%Y-%b-%d %H:%M:%S'`.

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
     info = TimeInfo()
 
+    @classmethod
+    def strptime(cls, time_string, format_string):
```
Make the call signature be:
```
def strptime(cls, format_spec, val, **kwargs):
```
Then pass `**kwargs` to the `cls` initializer later.  This allows specifying a different scale, precision, etc.

You will have to handle `format` specially if the user passes that in `kwargs`.  Something like:
```
format = kwargs.pop('format', None)
out = cls(*iterator.operands[1:], format='isot', **kwargs)
if format is not None:
    out.format = format

return out
```

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
+        to_string = (str if time_array.dtype.kind == 'U' else
+                     lambda x: str(x.item(), encoding='ascii'))
+        iterator = np.nditer([time_array, None],
+                             op_dtypes=[time_array.dtype, '<U30'])
```
Use `'U30'` not `'<U30'`.

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
+
+        Returns
+        -------
+        astropy.time.Time
```
Should be:
```
time_obj : `~astropy.time.Time`
    New `~astropy.time.Time` object corresponding to the input ``val``.
```

taldcroft commented on astropy/time/tests/test_basic.py:
```diff
@@ -1308,3 +1308,57 @@ def test_setitem_deltas():
+                        ['1998-01-01 00:00:03', '1998-01-01 00:00:04']])
+    t = Time.strptime(tstrings, '%Y-%b-%d %H:%M:%S')
+
+    assert np.all(t == time_object)
```
Test that the output `shape` is correct.

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
+        """
+        time_array = np.array(time_string)
+        to_string = (str if time_array.dtype.kind == 'U' else
+                     lambda x: str(x.item(), encoding='ascii'))
```
This part needs to be tested by passing byte strings.

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
+
+        """
+        time_array = np.array(time_string)
+        to_string = (str if time_array.dtype.kind == 'U' else
```
Check that `val.dtype.kind in ('U', 's')`, and if not raise a `TypeError` exception that the input needs to be strings.  Add a corresponding test of that exception being raised for bad input.

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
+            A new `Time` object corresponding to the input time_string.
+
+        """
+        time_array = np.array(time_string)
```
Write this as an in-place transformation with no copy:
```
val = np.asarray(val)
```

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,40 @@ def now(cls):
+    def strptime(cls, time_string, format_string):
+        """
+        Parse a string to a Time according to a format specification.
+        See time.strptime documentation for format specification.
```
Link to `time.strptime` docs (see similar comment for `strftime` below).

vn-ki commented:
I'm supposed add the doc [in this file](https://github.com/astropy/astropy/blob/master/docs/time/index.rst#timezones), right?

vn-ki commented:
@taldcroft `time.strptime` does not support `%f` because `struct_time` does not have milliseconds.

What should we do about it? Should we 

> Assume that anything following "." on the right side is a floating fraction of a second

like we assume in `TimeString` in formats.py or is there a better method?

Maybe we could use `datetime.strptime` if there is `%f` in the format string.

Cadair commented:
@taldcroft can you elaborate a little on this comment:

> Using `__format__` is fine, but it suffers from using a syntax language that is completely different from normal Python formatting and is really specific to strftime.

I was thinking that something along the lines of how `datetime.datetime` works with format would be appropriate here. Basically just:

```python
def __format__(self, format_spec):
    return self.strftime(format_spec)
```

(Maybe with some error checking) this would allow `Time` and `datetime` to be reasonably interchangeable as far as use through `.format()` goes.

taldcroft commented:
Here is my first reaction, which is not set completely in stone so if you want to push back then go for it.  :smile:

@Cadair - I'm a bit worried about locking in to being able to pass a [strftime format specifier](https://docs.python.org/3/library/time.html#time.strftime) into the generic `__format__` function because then you can get stuck there having to distinguish between a `strftime` format and something else like `20.8f` (which could be useful for numerical formats like `jd`).

In other words, `strftime` defines a well-known formatting language, which is one nice way to represent datetimes subject to certain limitations.  If you want to use that, then do so with the `strftime()` method that has the exact same API as `datetime.strftime()`.

At some level I think doing it this way makes code more explicit, though admittedly longer.
```
tm = Time('2001-01-10')
out = 'The time is {}'.format(tm.strftime("%b %d %Y %H:%M:%S"))
# vs.
out = 'The time is {:%b %d %Y %H:%M:%S}'.format(tm)
```
Honestly if I saw the second my first reaction would be a what the ...?  But the first example gives a nice reminder that this is a strftime format, and googling will quickly get me to the string format language definition.

taldcroft commented on astropy/time/core.py:
```diff
@@ -10,7 +10,8 @@
 import operator
-from datetime import datetime
+from datetime import datetime, date
+from time import strftime, strptime
```
Seems that this be `datetime` instead of `time`.  The `time` object doesn't support fractional seconds and no `%f` format option.

vn-ki replied:
`datetime` doesnot support leap seconds. If we use time, I think we are presenting the users with a sensible choice. If you want to use `strftime` without `%f` but with leap seconds, use `Time.strftime`. If you want `%f` and not leap seconds, use `Time.datetime.strftime`

taldcroft replied:
Ugh, OK.  Then never mind the comments about using `datetime`.

taldcroft replied:
I edited my comments accordingly.

taldcroft commented on docs/time/index.rst:
```diff
@@ -1270,6 +1270,29 @@ for example, may be more convenient to use to create `~datetime.tzinfo`
 
+strftime and strptime
+---------------------
+Sometimes you want the time represented by a `~astropy.time.Time` object to be custom
```
Replace this text with:
```
The `~astropy.time.Time` object supports output string representation
using the `format specification language
<https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior>`_
defined in the Python standard library for `datetime.strftime`.
This can be done using the `~astropy.time.Time.strftime`
method::
```

taldcroft commented on docs/time/index.rst:
```diff
@@ -1270,6 +1270,29 @@ for example, may be more convenient to use to create `~datetime.tzinfo`
+  >>> t.strftime('%H:%M:%S %d %b %Y')
+  '10:12:58 01 Jan 2018'
+
+Further examples and documentation for the format specification supported can
```
Cut this (links already provided above).

taldcroft commented on docs/time/index.rst:
```diff
@@ -1270,6 +1270,29 @@ for example, may be more convenient to use to create `~datetime.tzinfo`
+Further examples and documentation for the format specification supported can
+be found in the documentation of python standard library `~time.strftime`.
+
+If you want to get a `~astropy.time.Time` object from a string but don't find the
```
Replace with:
```
Conversely, to create a `~astropy.time.Time` object from a custom date string
that can be parsed with Python standard library `datetime.strptime`
(using the same format language linked above), use the
`~astropy.time.Time.strptime` class method::
```

taldcroft commented on astropy/time/core.py:
```diff
@@ -464,6 +514,37 @@ def __repr__(self):
+            datetime_tuple = (sk['year'], sk['mon'], sk['day'],
+                              sk['hour'], sk['min'], sk['sec'],
+                              date_tuple[6], date_tuple[7], -1)
+            formatted_strings.append(strftime(format_spec, datetime_tuple))
```
Better to `import datetime` at the top and refer to this (and `strptime`) explicitly as `datetime.strftime`.

taldcroft commented on astropy/time/core.py:
```diff
@@ -408,6 +409,55 @@ def now(cls):
+                             op_dtypes=[time_array.dtype, 'U30'])
+
+        for time, formatted in iterator:
+            time_tuple = strptime(to_string(time), format_string)
```
Use `datetime.strptime` to be explicit here.

taldcroft commented on astropy/time/tests/test_basic.py:
```diff
@@ -1308,3 +1308,105 @@ def test_setitem_deltas():
     assert not hasattr(t, '_delta_ut1_utc')
+
+
+def test_strftime_scalar():
```
Also test the `%f` microseconds output somewhere.

taldcroft commented on docs/time/index.rst:
```diff
@@ -1278,22 +1278,24 @@ for example, may be more convenient to use to create `~datetime.tzinfo`
-Sometimes you want the time represented by a `~astropy.time.Time` object to be custom
-formatted but don't find the need for a custom format, then you can use
-`~astropy.time.Time.strftime`::
+Custom string formats with `strftime` and `strptime`
```
These need to be double-backticked
```
Custom string formats with ``strftime`` and ``strptime``
```

taldcroft commented:
And for the next commit do not `[skip ci]` to make sure that the docs are OK.

vn-ki commented:
@taldcroft We can have `%f` support with strptime and strftime if we leverage the private API of python `_strptime` (Kudos @Cadair for finding this). 

If you are okay with it, I can try implementing that here.

taldcroft commented:
@vn-ki @Cadair - hmm that seems just a little dodgy.  Do you understand why that functionality isn't exposed in the public API?

taldcroft commented:
Assuming the astropy API here won't change very much, it might be best to make the one doc fix and get this merged as-is.  Then start a new PR which makes that change which we can discuss on its own merits.

vn-ki commented:
@taldcroft Done. Also I have added a fix for not parsing 3 digit years correctly.

vn-ki commented:
@taldcroft Any more comments here?

taldcroft commented:
@vn-ki - sorry, dropped this and forgot to see before that a changelog entry is needed.  You should put this as a new feature for 3.1 in the astropy.time section.  Otherwise all good.

taldcroft commented on CHANGES.rst:
```diff
@@ -129,6 +129,8 @@ astropy.time
 
 - Added `datetime.timedelta` format class for ``TimeDelta``. [#7441]
 
+- Added `strftime` and `strptime` methods to ``Time`` class. [#7323]
```
Both `strftime` and `strptime` should use double backticks.  Also give the readers some context for why they should care:
```
These methods are similar to those in the Python standard library 
`time` package and provide flexible input and output formatting.
```

vn-ki replied:
I do not seem to get the hang of back ticks. It should be fixed now.

vn-ki commented:
@taldcroft Done and rebased.

vn-ki commented:
Gentle ping @taldcroft. 

The gsoc coding period is coming to an end (ends in less than 2 weeks). If I could get this merged and the next PR by then, it would be awesome. Can I get a final review on this?

taldcroft commented:
@vn-ki - looks great now, sorry for the delay!  Thanks for the nice contribution, merging now.

vn-ki commented:
Thanks @taldcroft
